### PR TITLE
IRGen: Note use of types in conditional requirements of protocol conformances

### PR DESCRIPTION
**Pull request by aschwaighofer**
It is possible for the only mention of a conditional requirement to appear in
the protocol conformance record which is too late for lazy metadata type
emission.

SR-12891
rdar://63819461

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -2114,6 +2114,46 @@ static bool isConstantWitnessTable(SILWitnessTable *wt) {
   return true;
 }
 
+static void noteUseOfTypeMetadataInConditionalRequirements(
```
Comment (slavapestov) [inline lib/IRGen/GenProto.cpp:2117]: Why isn't this handled by emitting those mangled names in the witness table?

Comment (aschwaighofer) [reply lib/IRGen/GenProto.cpp:2117]: This is the witness table:

```
private enum E {
}

private protocol P {
  associatedtype AT
}

private struct S<A, B> {
  init()
}

extension S : P where A == E {
  typealias AT = B
}
sil_witness_table private <A, B where A == E> S<A, B>: P module ProtoCrash {
  associated_type AT: B
}
```

Which mangled names are you referring to?

Comment (aschwaighofer) [reply lib/IRGen/GenProto.cpp:2117]: Here is the ordering of code:
```
    // Okay, emit any definitions that we suddenly need.                        
    irgen.emitLazyDefinitions();                                                
                                                                                
    // Register our info with the runtime if needed.                            
    if (Opts.UseJIT) {                                                          
      IGM.emitBuiltinReflectionMetadata();                                      
      IGM.emitRuntimeRegistration();                                            
    } else {                                                                    
      // Emit protocol conformances into a section we can recognize at runtime. 
      // In JIT mode these are manually registered above.                       
      IGM.emitSwiftProtocols();                                                 
      IGM.emitProtocolConformances();
```

IGM.emitProtocolConformances will create typedefs (which note the use of type metadata) but this is too late.

Comment (slavapestov) [reply lib/IRGen/GenProto.cpp:2117]: Does it work if you emit the lazy definitions after protocol conformances?

We might need a loop to pump this until it stops adding new lazy definitions. I'm thinking of a case where emitting conditional requirements triggers lazy metadata emission, which triggers emitting another witness table, and so on.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test windows

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compatibility

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Pull request merged**